docs/graph.md

```diff
@@ -0,0 +1,714 @@
+    * Lastly, (and at the risk of torturing this metaphor), if you're a fan of medieval tools like mallets and untyped Python, you probably won't like nail guns or our approach to graphs. (But then again, if you're not a fan of type hints in Python, you've probably already bounced off PydanticAI to use one of the toy agent frameworks — good luck, and feel free to borrow my sledgehammer when you realize you need it)
+
+    In short, graphs are a powerful tool, but they're not the right tool for every job. Please consider other [multi-agent approaches](multi-agent-applications.md) before proceeding.
+
+    Unless you're sure you need a graph, you probably don't.
+
+Graphs and finite state machines (FSMs) are a powerful abstraction to model, control and visualize complex workflows.
+
+Alongside PydanticAI, we've developed `pydantic-graph` — an async graph and state machine library for Python where nodes and edges are defined using type hints.
+
+While this library is developed as part of the PydanticAI; it has no dependency on `pydantic-ai` and can be considered as a pure graph library. You may find it useful whether or not you're using PydanticAI or even building with GenAI.
+
+`pydantic-graph` is designed for advanced users and makes heavy use of Python generics and types hints. It is not designed to be as beginner-friendly as PydanticAI.
+
+!!! note "Very Early beta"
+    Graph support was [introduced](https://github.com/pydantic/pydantic-ai/pull/528) in v0.0.19 and is in very earlier beta. The API is subject to change. The documentation is incomplete. The implementation is incomplete.
+
+## Installation
+
+`pydantic-graph` is a required dependency of `pydantic-ai`, and an optional dependency of `pydantic-ai-slim`, see [installation instructions](install.md#slim-install) for more information. You can also install it directly:
+
+```bash
+pip/uv-add pydantic-graph
+```
+
+## Graph Types
+
+Graphs are made up of a few key components:
+
+### GraphContext
+
+[`GraphContext`][pydantic_graph.nodes.GraphContext] — The context for the graph run, similar to PydanticAI's [`RunContext`][pydantic_ai.tools.RunContext], this holds the state of the graph and is passed to nodes when they're run.
+
+`GraphContext` is generic in the state type of the graph it's used in, [`StateT`][pydantic_graph.state.StateT].
+
+### End
+
+[`End`][pydantic_graph.nodes.End] — return value to indicates the graph run should end.
+
+`End` is generic in the graph return type of the graph it's used in, [`RunEndT`][pydantic_graph.nodes.RunEndT].
+
+### Nodes
+
+Subclasses of [`BaseNode`][pydantic_graph.nodes.BaseNode] to define nodes.
+
+Nodes which are generally [`dataclass`es][dataclasses.dataclass] include:
+
+* any parameters required when calling the node
+* the business logic to execute the node
+* return annotations which are read by `pydantic-graph` to determine the outgoing edges of the node
```

Comment:
```suggestion
* return annotations, which are read by `pydantic-graph` to determine the outgoing edges of the node
```
(might make sense to say which method's return annotations will be used for this; it strikes me as weird to read "return annotations" on its own in the description of a class.)